### PR TITLE
feat: partition table query optimize

### LINKED ISSUE #1441: Query improvement about the query partition table
### Describe This Problem

Considering the query targeting at a partition table whose hash partition key is called `partition_col`:
`select * from partition_table where partition_col in ("a", "b", "c", ...)`.

And all the sub query plans share the same predicate, and if the inlist is large, the `min-max` and `bloom-filter` index may exhibit a very bad performance. However, actually, most of the values in the inlist don't exist at one specific partition, that is to say, the predicate in the sub query plan can be simplified into a more simple one.

### Proposal

Introduce an optimization procedure to remove the unnecessary values in the in-list predicate of the distributed sub query plan.

More implementation details are necessary before coding.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by zealchen**
## Rationale
Close #1441 

### TLDR
The performance issue with inlist queries is due to the extra overhead from bloom-filter-like directory lookups when scanning each SST file for rows. The solution is to create a separate predicate for each partition, containing only the keys relevant to that partition. Since the current partition filter only supports BinaryExpr(Column, operator, Literal) and non-negated InList expressions, this solution will address only those specific cases.

### Changes
1. During the scan building process, when identifying the partitions for a query, we create a PartitionedFilterKeyIndex variable to store the predicate key indices for each expression.
2. In the compute_partition_for_keys_group function, we use a HashMap<partition_id, HashMap<filter_index, BTreeSet<key_index>>> to record the indices of keys involved in partition computation for each group. 
3. In the partitioned_predicates function, we construct the final predicates for each partition.
4. In resolve_partitioned_scan_internal, we generate separate requests for each partition.

e.g.
conditions:
1. table schema: col_ts, col1, col2, in which col1 and col2 are both keys,
     and with two partitions
2. sql: select * from table where col1 = '33' and col2 in ("aa", "bb",
     "cc", "dd")

partition expectations:
   yield two predicates
      p0: col1 = '33' and col2 in ("aa", "bb", "cc");
      p1: col1 = '33' and col2 in ("dd")

### Other issues discovered
When the inlist key args length is less than three, Expr will be refactored to nested BinaryExpr which bypasses the FilterExtractor. 

e.g.
SQL: select * from table where col1 in ("aa", "bb") and col2 in (1,2,3,4,5...1000)
Since ("aa", "bb") has fewer than three elements, the col1 key filter is not included in partition computation, which interrupts the partitioning process in the get_candidate_partition_keys_groups function, as contains_empty_filter is set to true.

## Test Plan
1. UT: test_partitioned_predicate
2. Manual test.

### CODE REVIEW

File: src/df_engine_extensions/src/dist_sql_query/resolver.rs [hunk 1]
```diff
@@ -135,10 +135,17 @@ impl Resolver {
...
                 .into_iter()
-                .map(|table| {
+                .enumerate()
+                .map(|(idx, table)| {
                     let plan = Arc::new(UnresolvedSubTableScan {
                         table: table.clone(),
-                        table_scan_ctx: unresolved.table_scan_ctx.clone(),
+                        table_scan_ctx: if let Some(ref predicates) = unresolved.predicates {
+                            let mut ctx = unresolved.table_scan_ctx.clone();
+                            ctx.predicate = Arc::new(predicates[idx].clone());
```
Comment (jiacai2050) [inline src/df_engine_extensions/src/dist_sql_query/resolver.rs:144]: Add comments why we need to overwrite old predicate.

File: src/partition_table_engine/src/scan_builder.rs [hunk 2]
```diff
@@ -89,18 +92,186 @@ impl TableScanBuilder for PartitionedTableScanBuilder {
...
+            self.get_sub_table_idents(&self.table_name, &self.partition_info, &partitions);
+
+        let predicates = if partitioned_key_indices.len() == partitions.len() {
+            Some(
+                partitioned_predicates(
+                    request.predicate.clone(),
+                    &partitions,
+                    &mut partitioned_key_indices,
+                )
+                .map_err(|e| DataFusionError::Internal(format!("err:{e}")))?,
```
Comment (jiacai2050) [inline src/partition_table_engine/src/scan_builder.rs:113]:
```suggestion
                .map_err(|e| DataFusionError::Internal(format!("partition predicates failed, err:{e}")))?,
```

File: src/partition_table_engine/src/scan_builder.rs [hunk 3]
```diff
@@ -89,18 +92,186 @@ impl TableScanBuilder for PartitionedTableScanBuilder {
...
+                .unwrap()
+                .primary_key_indexes(vec![1, 2])
+                .build()
+                .unwrap()
+        };
+
+        let df_partition_rule = DfPartitionRuleAdapter::new(partition_info, &schema).unwrap();
+
+        let exprs = vec![
+            Expr::BinaryExpr(BinaryExpr {
```
Comment (jiacai2050) [inline src/partition_table_engine/src/scan_builder.rs:203]: DataFusion provider some helper functions to define expr:
- https://docs.rs/datafusion/latest/datafusion/logical_expr/index.html#functions

File: src/table_engine/src/partition/rule/key.rs [hunk 4]
```diff
@@ -139,15 +141,33 @@ impl KeyRule {
         &self,
         group: &[usize],
         filters: &[PartitionFilter],
-    ) -> Result<BTreeSet<usize>> {
+    ) -> Result<PartitionedFilterKeyIndex> {
         let mut partitions = BTreeSet::new();
+        // Retrieve all the key DatumView instances along with their corresponding
+        // indices related to their positions in the predicate inlist. Since DatumView
```
[the remark this answers is not in the document]
Comment (zealchen) [reply src/table_engine/src/partition/rule/key.rs:147]: Only the Hash trait is not adequate, since a HashSet requires that the elements implement the [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html) and [Hash](https://doc.rust-lang.org/std/hash/trait.Hash.html) traits.

[the remark this answers is not in the document]
Comment (zealchen) [reply src/table_engine/src/partition/rule/key.rs:147]: This BTreeSet in PartitionedFilterKeyIndex is used to deduplicate the inlist key value that contributes to the partition calculation.

File: integration_tests/cases/env/cluster/ddl/partition_table.sql [hunk 5]
```diff
@@ -57,15 +57,15 @@ SELECT * from partition_table_t where name in ("horaedb5", "horaedb6", "horaedb7
...
--- SQLNESS REPLACE metrics=\[.*?s\] metrics=xx
+-- SQLNESS REPLACE elapsed_compute=\d+.?\d*(µ|m|n) elapsed_compute=xx
 -- SQLNESS REPLACE scan_memtable_\d+ scan_memtable_n
 EXPLAIN ANALYZE SELECT * from partition_table_t where name = "ceresdb0";
 
 -- SQLNESS REPLACE duration=\d+.?\d*(µ|m|n) duration=xx
 -- SQLNESS REPLACE compute=\d+.?\d*(µ|m|n) compute=xx
 -- SQLNESS REPLACE __partition_table_t_\d __partition_table_t_x
 -- SQLNESS REPLACE time=\d+.?\d*(µ|m|n) time=xx
--- SQLNESS REPLACE metrics=\[.*?s\] metrics=xx
```
Comment (zealchen) [inline integration_tests/cases/env/cluster/ddl/partition_table.sql:68]: This regular expression was used to replace duration value in metrics (metrics=[output_rows=0, elapsed_compute=15.5µs]). instead, it replaces more text which causing overwriting predicate metrics.

Orignal text:
plan_type,plan,\nString(\"Plan with Metrics\"),String(\"ResolvedPartitionedScan: pushdown_continue:false, partition_count:3, metrics=[\\npartition_table_t:\\n    __partition_table_t_0:\\n        poll_duration=65.875µs\\n        total_duration=1.732042ms\\n        wait_duration=1.666167ms\\n    __partition_table_t_1:\\n        poll_duration=44.5µs\\n        total_duration=2.275958ms\\n        wait_duration=2.231458ms\\n    __partition_table_t_3:\\n        poll_duration=69.249µs\\n        total_duration=2.367166ms\\n        wait_duration=2.297917ms\\n\\n__partition_table_t_0:\\nCoalescePartitionsExec, metrics=[output_rows=0, elapsed_compute=21.5µs]\\n  ScanTable: table=__partition_table_t_0, parallelism=8, priority=Low, partition_count=UnknownPartitioning(8), metrics=[\\nPredicate { exprs:[name IN ([Utf8(\\\"ceresdb2\\\")])], time_range:TimeRange { inclusive_start: Timestamp(-9223372036854775808), exclusive_end: Timestamp(9223372036854775807) } }\\nscan_table:\\n    do_merge_sort=true\\n    iter_num=1\\n    merge_iter_0:\\n        init_duration=162.042µs\\n        num_memtables=0\\n        num_ssts=0\\n        scan_count=1\\n        scan_duration=0ns\\n        times_fetch_row_from_multiple=0\\n        times_fetch_rows_from_one=0\\n        total_rows_fetch_from_one=0\\n        scan_memtable_28, fetched_columns:[tsid,t,name,id,value]:\\n=0]\\n\\n__partition_table_t_1:\\nCoalescePartitionsExec, metrics=[output_rows=0, elapsed_compute=24.791µs]\\n  ScanTable: table=__partition_table_t_1, parallelism=8, priority=Low, partition_count=UnknownPartitioning(8), metrics=[\\nPredicate { exprs:[name IN ([Utf8(\\\"ceresdb0\\\"), Utf8(\\\"ceresdb4\\\")])], time_range:TimeRange { inclusive_start: Timestamp(-9223372036854775808), exclusive_end: Timestamp(9223372036854775807) } }\\nscan_table:\\n    do_merge_sort=true\\n    iter_num=1\\n    merge_iter_0:\\n        init_duration=124.875µs\\n        num_memtables=0\\n        num_ssts=0\\n        scan_count=1\\n        scan_duration=42ns\\n        times_fetch_row_from_multiple=0\\n        times_fetch_rows_from_one=0\\n        total_rows_fetch_from_one=0\\n        scan_memtable_5, fetched_columns:[tsid,t,name,id,value]:\\n=0]\\n\\n__partition_table_t_3:\\nCoalescePartitionsExec, metrics=[output_rows=0, elapsed_compute=20.709µs]\\n  ScanTable: table=__partition_table_t_3, parallelism=8, priority=Low, partition_count=UnknownPartitioning(8), metrics=[\\nPredicate { exprs:[name IN ([Utf8(\\\"ceresdb1\\\"), Utf8(\\\"ceresdb3\\\")])], time_range:TimeRange { inclusive_start: Timestamp(-9223372036854775808), exclusive_end: Timestamp(9223372036854775807) } }\\nscan_table:\\n    do_merge_sort=true\\n    iter_num=1\\n    merge_iter_0:\\n        init_duration=162.458µs\\n        num_memtables=0\\n        num_ssts=0\\n        scan_count=1\\n        scan_duration=42ns\\n        times_fetch_row_from_multiple=0\\n        times_fetch_rows_from_one=0\\n        total_rows_fetch_from_one=0\\n        scan_memtable_1, fetched_columns:[tsid,t,name,id,value]:\\n=0]\\n=0]\\n\"),\n

Comment (jiacai2050) [reply integration_tests/cases/env/cluster/ddl/partition_table.sql:68]: Good catch!

**Review by jiacai2050 — APPROVED** [review]
LGTM

File: src/table_engine/src/partition/rule/df_adapter/mod.rs [hunk 6]
```diff
@@ -16,21 +16,30 @@
...
-        PartitionedRows,
+        df_adapter::extractor::FilterExtractorRef, factory::PartitionRuleFactory,
+        filter::PartitionFilter, PartitionRulePtr, PartitionedRows,
     },
     BuildPartitionRule, PartitionInfo, Result,
 };
 
 mod extractor;
 
+pub type PartitionId = usize; // partiton number (id)
```
Comment (jiacai2050) [inline src/table_engine/src/partition/rule/df_adapter/mod.rs:35]: 👍 for comments
